text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+The `base_image` arg allows the Dockerfile to reference the original base image.
+The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 
+When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layers will be rebuilt on the next build (as the value will change).
+
+Build args specified in `build.toml` are provided to `build.Dockerfile` or `Dockerfile` (when applied to the build-time base image).
+Build args specified in `launch.toml` are provided to `run.Dockerfile` or `Dockerfile` (when applied to the runtime base image).
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically on the base image before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. If multiple Dockerfiles are applied, all must set `io.buildpacks.rebasable=true` for the final value to be `true`. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+#### Example: App-specified Dockerfile Extension
+
+This example extension would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/ext/com.example.appext/bin/build`
+```
+#!/bin/sh
+[ -f build.Dockerfile ] && cp build.Dockerfile "$1/"
+[ -f run.Dockerfile ] && cp run.Dockerfile "$1/"
+```
+
+#### Example: RPM Dockerfile Extension (app-based)
+
+This example extension would allow a builder to install RPMs for each language runtime, based on the app directory.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+##### `/cnb/ext/com.example.rpmext/bin/build`
+```
+#!/bin/sh
+[ -f Gemfile.lock ] && cp "$CNB_BUILDPACK_DIR/Dockerfile-ruby" "$1/Dockerfile"
+[ -f package.json ] && cp "$CNB_BUILDPACK_DIR/Dockerfile-node" "$1/Dockerfile"
+```
+
+
+### Dockerfiles for Base Images
+
+The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.
+
+Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
```

Comment:
A discussion and RFC/WIP has also been created on opencontainer to support SBOM part of an image METADATA using a blob store json object - see: https://github.com/opencontainers/artifacts/pull/37
So instead of putting the SBOM json file generated as `io.buildpacks.sbom` we could rely on a more generic approach relying on image METADATA. WDYT ? @sclevine

Reply:
I agree that `io.buildpacks.sbom` is not a long term solution. I expected https://github.com/buildpacks/rfcs/pull/166 to specify a more complete solution that included both the stack and buildpack SBoM. It looks like the base image SBoM was de-scoped from #166, so we need a follow-up RFC to replace `io.buildpacks.sbom` and specify that the base image SBoM should be merged into the final SBoM. 

Until https://github.com/opencontainers/artifacts/pull/37 lands and is widely supported in registries, I think we should use cosign's SBoM support: https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md

But before we implement either of those standardized solutions, maybe we should move the stack SBoM to an FS layer and ensure that it's merged into SBoM generated in #166?

Reply:
> Until [opencontainers/artifacts#37](https://github.com/opencontainers/artifacts/pull/37) lands and is widely supported in registries, I think we should use cosign's SBoM support: https://github.com/sigstore/cosign/blob/main/specs/SBOM_SPEC.md

That will request additional work then to migrate from cosign to opencontainer spec when we will move and to implement a mechanism able to convert cosign SBOM to opencontainer SBOM ...

Reply:
> That will request additional work then to migrate from cosign to opencontainer spec when we will move and to implement a mechanism able to convert cosign SBOM to opencontainer SBOM ...

What is the alternative? I don't think we should implement the OCI version until it is merged and implemented in popular registries, which I don't expect to happen in the near future.

(And as @dlorenc mentions, it's just the storage mechanism, not the format of the SBoM itself.)

Reply:
Can we clarify the interplay between `genpkgs` and any sbom files output by hooks? It's kind of confusing that both mechanisms are allowed. When would you use one vs the other?

Reply:
Marking as resolved given that we're leaving SBOM generation up to the platform (see https://github.com/buildpacks/spec/pull/298#issuecomment-1099543096).